tests/test_generate_data.py

```diff
@@ -3,12 +3,464 @@
+                        + "\n"
+                        + seed_example["context"].strip(),
+                        "assistant": qna["answer"].strip(),
+                    }
+                )
+
+        else:
+            # FIXME: handle freeform skills - no context
+            test_samples.append(
+                {
+                    "user": seed_example["question"] + "\n" + seed_example["context"],
+                    "assistant": seed_example["answer"],
+                }
+            )
+    return test_samples
+
+
+def generate_train_samples(yaml_contents):
+    """Generate expected training samples in the user/assistant format
+    used by the legacy training methods such as Linux training and MacOS
+    training.
+
+    Mirroring _noop_llmblock_generate() below, we generate 10 samples
+    per input, and then follow _gen_train_data()'s output format.
+    """
+
+    def add_q(q):
+        return (q + "?") if not "?" in q else q
+
+    train_samples = []
+    is_knowledge = "document" in yaml_contents
+    for seed_example in yaml_contents["seed_examples"]:
+        for i in range(10):
+            if is_knowledge:
+                train_samples.append(
+                    {
+                        "user": seed_example["context"]
+                        + f" (q{i}) "
+                        + add_q(
+                            seed_example["questions_and_answers"][0]["question"].strip()
+                        ),
+                        "assistant": f"(a{i}) "
+                        + seed_example["questions_and_answers"][0]["answer"].strip(),
+                    }
+                )
+            else:
+                # FIXME: handle freeform skills - no context
+                train_samples.append(
+                    {
+                        "user": seed_example["context"]
+                        + f" (q{i}) "
+                        + add_q(seed_example["question"]),
+                        "assistant": f"(a{i}) " + seed_example["answer"],
+                    }
+                )
+    return train_samples
+
+
+def load_test_skills(skills_file_path) -> Union[Dict[str, Any], None]:
+    with open(skills_file_path, "r", encoding="utf-8") as skills_file:
+        return yaml.safe_load(skills_file)
+
+
+def _noop_llmblock_generate(self, samples):
+    """Generate mock output based on input samples.
+
+    Simply return the seed question and response from the input sample,
+    joined using '?' and with an integer discriminator.
+
+    _get_question_hack() and _get_response_hack() is the code that later
+    splits these using the '?' separator.
+
+    Return 10 output samples per input samples, since the LLMBlock in the
+    simple pipeline is configured with 'n: scaled' and we pass
+    num_instructions_to_generate=10 to generate_data.
+    """
+
+    def strip_q(q):
+        return q.strip().rstrip("?")
+
+    output = []
+    for sample in samples:
+        for i in range(10):
```

Comment:
same question about the 10. Maybe this is best stored as a global variable that each of these loops can use so as to avoid duplication.

Reply:
Good point. I'll add a constant to be used.

In this case, we generate 10 output samples per input samples, since the `LLMBlock` in the simple pipeline is configured with 'n: scaled' and we pass `num_instructions_to_generate=10` to `generate_data()`.